frostsnap_core/src/device.rs

```diff
@@ -75,19 +76,120 @@ pub struct AccessStructureData {
 pub struct EncryptedSecretShare {
     /// The image of the encrypted secret. The device stores this so it can tell the coordinator
     /// about it as part of the recovery system.
-    pub image: Point<Normal, Public, Zero>,
+    pub share_image: ShareImage,
     /// The encrypted secret share
     pub ciphertext: Ciphertext<32, Scalar<Secret, Zero>>,
 }
 
+impl EncryptedSecretShare {
+    pub fn encrypt(
+        secret_share: SecretShare,
+        access_structure_ref: AccessStructureRef,
+        coord_contrib: CoordShareDecryptionContrib,
+        symm_keygen: &mut impl DeviceSymmetricKeyGen,
+        rng: &mut impl rand_core::RngCore,
+    ) -> Self {
+        let share_image = ShareImage::from_secret(secret_share);
+        let encryption_key = symm_keygen.get_share_encryption_key(
+            access_structure_ref,
+            secret_share.index,
+            coord_contrib,
+        );
+        let ciphertext = Ciphertext::encrypt(encryption_key, &secret_share.share, rng);
+        EncryptedSecretShare {
+            share_image,
+            ciphertext,
+        }
+    }
+
+    pub fn decrypt(
+        &self,
+        access_structure_ref: AccessStructureRef,
+        coord_contrib: CoordShareDecryptionContrib,
+        symm_keygen: &mut impl DeviceSymmetricKeyGen,
+    ) -> Option<SecretShare> {
+        let encryption_key = symm_keygen.get_share_encryption_key(
+            access_structure_ref,
+            self.share_image.share_index,
+            coord_contrib,
+        );
+
+        self.ciphertext
+            .decrypt(encryption_key)
+            .map(|share| SecretShare {
+                index: self.share_image.share_index,
+                share,
+            })
+    }
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct KeyGenPhase1 {
+    pub device_to_share_index: BTreeMap<DeviceId, NonZeroU32>,
+    pub input_state: encpedpop::Contributor,
+    pub threshold: u16,
+    pub key_name: String,
+    pub key_purpose: KeyPurpose,
+}
+
+#[derive(Clone, Debug)]
+pub struct KeyGenPhase2 {
+    pub keygen_id: KeygenId,
+    secret_share: PairedSecretShare,
+    agg_input: encpedpop::AggKeygenInput,
+    key_name: String,
+    key_purpose: KeyPurpose,
+}
+
+impl KeyGenPhase2 {
+    pub fn session_hash(&self) -> SessionHash {
+        SessionHash::from_agg_input(&self.agg_input)
+    }
+    pub fn key_name(&self) -> &str {
+        self.key_name.as_str()
+    }
+    pub fn t_of_n(&self) -> (u16, u16) {
+        (
+            self.agg_input.shared_key().threshold().try_into().unwrap(),
+            self.agg_input.encryption_keys().count().try_into().unwrap(),
+        )
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct BackupDisplayPhase {
```

Comment:
What makes this a "phase", because it will be used in other workflows like checking shares?

Reply:
Ah "Often confirmations contain what we call a "phase" which is both the data that describes the action and what will be passed back"